docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,200 @@
+for each required REST endpoint (belonging to another EdgeX service) directly from the associated *URL config key read from Consul (if
+enabled) or directly from the configuration file.
+
+The shared client initialization code also created an Endpoint monitor goroutine and passed it a go channel channel used by the service
+to receive updates to the REST API endpoint URL. This monitor goroutine effectively polled Consul every 15s (this became configurable in
+later versions) for the client's service address and if a change was detected, would write the updated endpoint URL to the given channel,
+effectively ensuring that the service started using the new URL.
+
+It wasn't till late in the Geneva development cycle that I noticed log messages which made me aware of the fact that every one of our
+services was making a REST call to check the address of a service endpoint every 15s, for **every** REST endpoint it used! An issue was
+filed (https://github.com/edgexfoundry/edgex-go/issues/2594), and the client monitoring was removed as part of the Geneva 1.2.1 release.
+
+## Problem Statement
+The fundamental problem with the existing implementations (as decribed above), is that there is too much duplication of configuration across
+services. For instance, Core Data's service port can easily be changed by passing the environment variable SERVICE_PORT to the service on
+startup. This overrides the configuration read from the configuration provider, and will cause Core Data to listen on the new port,
+however it has no impact on any services which use Core Data, as the client config for each is read from the configuration provider (excluding
+security-proxy-setup).
+
+This means in order to change a service port, environment variable overrides (e.g. CLIENTS_COREDARA_PORT) need to set for every client service as well as security-proxy-setup (if required).
+
+## Decision
+Update the core, support, and security-proxy-setup services to use go-mod-registry's ```Client.GetServiceEndpoint``` method on startup to determine (a) if
+a service dependency is available and (b) use the returned address information to initialize client endpoints (or setup the correct route in the
+case of proxy-setup). The same changes also need to be applied to the App Functions SDK and Go Device SDK, with only minor changes required in the C Device SDK (see previous commments re: the current implementation).
+
+*Note* - One impact of this decision is that as the security-proxy-setup service currently runs _before_ any of the core and support services are
+started, it would not be possible to implement this proposal without also modifying the service to use a lazy initialization of the API Gateway's
+routes. We should also explore whether it's possible to extend the registry abstraction to allow notification of new service registrations vs. forcing
+proxy-setup to poll for all known services.
```

Comment:
IMO this caveat about security-proxy-setup is fatal to applying this ADR to this service.  We need to define a mechanism defined that will keep security-proxy-setup from doing its thing until the required services are running, or figure out how to make proxy-setup dynamic in some way.  If we go ahead with this ADR as written proxy-setup will act non-deterministically.

Reply:
Define "lazy initialization". 

[**n00b dev**]  _**I may be totally wrong here**_  [/**n00b dev**], but it is my understanding that proxy setup doesn't wait for services to be up in order to create Kong services/routes/plugins - it just configures them through the Kong Admin API and exits. From my limited knowledge, the bootstrapping process proxy setup just waits on the Kong service and Kong DB to be up and running, and then the subsequent Go code starts executing. Yeah/no?

Reply:
> [**n00b dev**] _**I may be totally wrong here**_ [/**n00b dev**], but it is my understanding that proxy setup doesn't wait for services to be up in order to create Kong services/routes/plugins - it just configures them through the Kong Admin API and exits.

There is a fine line between a service REGISTERING ITSELF and the service BEING UP.  You're right.  Proxy setup doesn't require the service to be up.  But Tony's ADR requires the service to be REGISTERED in order to obtain the address and port for configuration.  The line is fine because the they between a service registering itself and being able to serve requests is only milliseconds.

So if we want to rely on the registry, we have a few options.

## Option 1

In Consul mode, proxy setup would use not its full compiled-in service registry, but simply a LIST OF REQUIRED SERVICES.  Proxy setup would either have to wait for all those services to have registered before configuring the proxy, or will need to be a long-running services that watches for new service registrations and adds them to them to the proxy if they appear in the list.

I am not aware of any use cases where Kong is used without Consul.


## Option 2

Leverage the ability to [attach metadata to service registrations](https://www.consul.io/api-docs/catalog#register-entity) and have each service add metadata that indicates that it wants to be exposed via the Kong proxy.  Proxy-setup could be a daemon process that, as above, watched for new service registrations with the flag set.  We could also encode metadata such as what ACL group(s?) the service belongs to, so that one could have finer-grained controls as to which services are available with a given Kong token.


The point I was trying to make above (the "fatal" bit), is that we can't hand-wave it.  It needs to be defined in the ADR.

Reply:
I think **Option 2** is doable. We can add a new property something  like `ExposeViaProxy` to the `RegistryInfo` configuration and update go-mod-register to add this to the metadata for service registry. The Registry config section would look like:
```
[Registry]
Host = 'localhost'
Port = 8500
Type = 'consul'
ExposeViaProxy = true
```
We would also expand the the Registry's interface to add an API to return all registered services and their metadata with optional metadata filter. 

I looked at the Consul API and verified that we can do the above.

proxy-setup would then be long lived adding proxy for any new services that are registered. Downside is it will have to poll the list of registered services periodically. 

 I looks like the same QueryOptions we use to wait for config changes exist on the API to get a Service's info (i.e. endpoint), but not on the API that returns all the registered services. This means that the Clients could use this to refresh the endpoint if it changes.